ts/session/sending/MessageQueue.ts

```diff
@@ -1,54 +1,167 @@
   }
-  public sendToGroup(message: ContentMessage | OpenGroupMessage) {
-    throw new Error('Method not implemented.');
+
+  public async sendMessageToDevices(
+    devices: Array<PubKey>,
+    message: ContentMessage
+  ) {
+    let currentDevices = [...devices];
+
+    // Sync to our devices if syncable
+    if (SyncMessageUtils.canSync(message)) {
+
+      const ourDevices = await SyncMessageUtils.getOurPairedDevices();
+      await this.sendSyncMessage(message, ourDevices);
+
+      // Remove our devices from currentDevices
+      const ourDeviceContacts = ourDevices.map(device => ConversationController.get(device.key));
+      currentDevices = _.xor(currentDevices, ourDeviceContacts);
+    }
+
+    const promises = currentDevices.map(async device => {
+      await this.queue(device, message);
+    });
+
+    return Promise.all(promises);
   }
-  public sendSyncMessage(message: ContentMessage) {
-    throw new Error('Method not implemented.');
+
+  public async sendToGroup(message: OpenGroupMessage | ContentMessage): Promise<boolean> {
+    if (
+      !(message instanceof OpenGroupMessage) &&
+      !(message instanceof ClosedGroupMessage)
+    ) {
+      return false;
+    }
+
+    // Closed groups
+    if (message instanceof ClosedGroupMessage) {
+      // Get devices in closed group
+      const conversation = ConversationController.get(message.groupId);
+      const recipientsModels = conversation.contactCollection.models;
+      const recipients: Array<PubKey> = recipientsModels.map(
+        (recipient: any) => new PubKey(recipient.id)
+      );
+
+      await this.sendMessageToDevices(recipients, message);
+
+      return true;
+    }
+
+    // Open groups
+    if (message instanceof OpenGroupMessage) {
+      // No queue needed for Open Groups; send directly
+      await MessageSender.sendToOpenGroup(message);
+
+      return true;
+    }
+
+    return false;
   }
 
-  public processPending(device: string) {
-    // TODO: implement
+  public async sendSyncMessage(
+    message: ContentMessage,
+    sendTo: Array<PubKey>
+  ) {
+    // Sync with our devices
+    const promises = sendTo.map(async device => {
+      const syncMessage = await SyncMessageUtils.from(message, device);
+
+      return this.queue(device, syncMessage);
+    });
+
+    return Promise.all(promises);
   }
 
-  private processAllPending() {
-    // TODO: Get all devices which are pending here
+  public async processPending(device: PubKey) {
+    const messages = this.pendingMessageCache.getForDevice(device);
+
+    const hasSession = SessionProtocol.hasSession(device);
+    const conversation = ConversationController.get(device.key);
+    const isMediumGroup = conversation.isMediumGroup();
+
+    if (!isMediumGroup && !hasSession) {
+      await SessionProtocol.sendSessionRequestIfNeeded(device);
+
+      return;
+    }
+
+    const jobQueue = this.getJobQueue(device);
+    messages.forEach(message => {
+      if (!jobQueue.has(message.identifier)) {
```

Comment:
`message.identifier` is not unique as we've stated ..
we should use `message.timestamp`, no need to also check for device as a `jobQueue` is per device.